File: ml/db/Models/DifficultyBanditInteractions.py
from datetime import datetime
from db import db
from db.Models.Base import Base
from pydantic import BaseModel, ConfigDict
from typing import Optional
import sqlalchemy as sa

engine = db.getEngine()


# DifficultyBanditInteractions Model
class DifficultyBanditInteractions(Base):
    __table__ = sa.Table(
        "difficulty_bandit_interactions", Base.metadata, autoload_with=engine
    )


# Pydantic model for DifficultyBanditInteractions
class DifficultyBanditInteractionsSchema(BaseModel):
    id: Optional[int] = None
    student_id: int
    skill_id: int
    item_id: int
    arm: str
    selection_source: str
    context: dict
    previous_mastery: float
    correct: bool
    new_mastery: float
    reward: float
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

File: src/app/Models/DifficultyBanditInteractions.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class DifficultyBanditInteractions extends Model
{
    protected $fillable = [
        'student_id',
        'skill_id',
        'question_id',
        'arm',
        'selection_source',
        'context',
        'previous_mastery',
        'is_correct',
        'new_mastery',
        'reward',
    ];

    public function student() {
        return $this->belongsTo(User::class);
    }

    public function skill() {
        return $this->belongsTo(Skills::class);
    }

    public function question() {
        return $this->belongsTo(Questions::class);
    }

    /**
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'context' => 'array',
        ];
    }
}
